README.md
<p align="center">
  <img width="420" src="assets/logo.png"/>
</p>

[![CocoaPods](https://cocoapod-badges.herokuapp.com/v/SubtleVolume/badge.svg)](http://cocoapods.org/?q=subtlevolume)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
![Swift 4.2](https://img.shields.io/badge/swift-4.2-orange.svg)

Replace the volume popup with a more subtle way to display the volume when the user changes it with the volume rocker.

<p align="center">
  <img width="640" src="assets/screenshot.png"/>
</p>

# Why and how
The iOS default popover showing the volume status that appears when the user clicks the volume rocker is a big obtrusive glossy view that covers the content shown. This library offers a way to show a more subtle indicator.  
To make sure that the popover is not shown there are two conditions that need to be satisfied:  
- An `AVAudioSession` needs to be active
- An `MPVolumeView` needs to be in the current view's hierarchy, and its alpha needs to be greater than 0

Once a `SubtleVolume` is added to your view, an audio session is automatically started, and the view's alpha is set to `0.0001` in the hidden state.

# Getting Started
Create an instance of `SubtleVolume` with one of its convenience initializers, and set its position (you can either set the frame or let autolayout handle it):
```swift
let volume = SubtleVolume(style: .plain)
volume.frame = CGRect(x: 0, y: 10, width: UIScreen.main.bounds.width, height: 4) // or wherever you like
```

Set the barTintColor property:
```swift
volume.barTintColor = .red
```

Set the animation type if needed (no animation will result in the indicator being always visible):
```swift
volume.animation = .slideDown
```

Add the view to your hierarchy:
```swift
view.addSubview(volume)
```

To change the volume programmatically:
```swift
try? volume.setVolumeLevel(0.5)
```

Or use the convenience methods:
```swift
try? volume.decreaseVolume(by: 0.2, animated: true)
try? volume.increaseVolume(by: 0.2, animated: true)
```

### Accessory image
You can provide an accessory image that will be shown to the bar's left. See the delegate method:
```swift
func subtleVolume(_ subtleVolume: SubtleVolume, accessoryFor value: Double) -> UIImage? {
  return value > 0 ? #imageLiteral(resourceName: "volume-on.pdf") : #imageLiteral(resourceName: "volume-off.pdf")
}
```

### iPhone X(S/R) support
Want to be fancy and show the volume bar in the notch area? Check out the demo project. The main gist is this:

```swift
class ViewController: UIViewController {
  let volume = SubtleVolume(style: .rounded)
  var statusBarVisible = true

  // ...

  override func viewDidLayoutSubviews() {
    super.viewDidLayoutSubviews()

    if view.safeAreaInsets.top > 0 {
      volume.padding = CGSize(width: 2, height: 8)
      volume.frame = CGRect(x: 16, y: 8, width: 60, height: 20)
    } else {
      // older phones here
    }
  }

  override var preferredStatusBarStyle: UIStatusBarStyle {
    return .lightContent
  }

  override var preferredStatusBarUpdateAnimation: UIStatusBarAnimation {
    return .slide
  }

  override var prefersStatusBarHidden: Bool {
    return !statusBarVisible
  }
}

extension ViewController: SubtleVolumeDelegate {
  func subtleVolume(_ subtleVolume: SubtleVolume, didChange value: Double) {
    if !subtleVolume.isAnimating && view.safeAreaInsets.top > 0 {
      statusBarVisible = true
      UIView.animate(withDuration: 0.1) {
        self.setNeedsStatusBarAppearanceUpdate()
      }
    }
  }

  func subtleVolume(_ subtleVolume: SubtleVolume, willChange value: Double) {
    if !subtleVolume.isAnimating && view.safeAreaInsets.top > 0 {
      statusBarVisible = false
      UIView.animate(withDuration: 0.1) {
        self.setNeedsStatusBarAppearanceUpdate()
      }
    }
  }
}
```


# Handle the background state

Once your app goes in background, you'll need to resume the session when it becomes active:

```swift
NotificationCenter.default.addObserver(volume, selector: #selector(SubtleVolume.resume), name: NSNotification.Name.UIApplicationDidBecomeActive, object: nil)
```

SubtleVolume automatically removes the observer on deinit.

# Hire us
Written by [Andrea Mazzini](https://twitter.com/theandreamazz). We're available for freelance work, feel free to us [here](https://www.fancypixel.it/contact/).

Want to support the development of [these free libraries](https://cocoapods.org/owners/734)? Buy me a coffee ☕️ via [Paypal](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=46FNZD4PDVNRU).  

# MIT License

	Copyright (c) 2017-2018 Andrea Mazzini. All rights reserved.

	Permission is hereby granted, free of charge, to any person obtaining a
	copy of this software and associated documentation files (the "Software"),
	to deal in the Software without restriction, including
	without limitation the rights to use, copy, modify, merge, publish,
	distribute, sublicense, and/or sell copies of the Software, and to
	permit persons to whom the Software is furnished to do so, subject to
	the following conditions:

	The above copyright notice and this permission notice shall be included
	in all copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
	OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
	MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
	IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
	CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
	TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
	SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
`SubtleVolume` adheres to [Semantic Versioning](http://semver.org/).

- `1.0.x` Releases - [1.0.0](#100) | [1.0.1](#101) | [1.0.2](#102)  
- `0.5.x` Releases - [0.5.0](#050) | [0.5.1](#051) | [0.5.2](#052)  
- `0.4.x` Releases - [0.4.0](#040)  
- `0.3.x` Releases - [0.3.0](#030)  
- `0.2.x` Releases - [0.2.0](#020) | [0.2.1](#021)  
- `0.1.x` Releases - [0.1.0](#010) | [0.1.1](#011)  

---

## [1.0.2](https://github.com/andreamazz/SubtleVolume/releases/tag/1.0.2)

- Fix possible state inconsistency

## [1.0.1](https://github.com/andreamazz/SubtleVolume/releases/tag/1.0.1)

- Fix delegate issue

## [1.0.0](https://github.com/andreamazz/SubtleVolume/releases/tag/1.0.0)

### Stable release
- Updated to Swift 4.2
- Improved support for iPhone X(S/R) screens
- Added accessory image

## [0.5.2](https://github.com/andreamazz/SubtleVolume/releases/tag/0.5.2)

- Merged #17  

## [0.5.1](https://github.com/andreamazz/SubtleVolume/releases/tag/0.5.1)

### Fixed

- Issue with `setVolumeLevel(_: animated:)` called with `animation` set to `false`

## [0.5.0](https://github.com/andreamazz/SubtleVolume/releases/tag/0.5.0)

### Added

- Volume convenience methods  
- Objective-C support

### Fixed

- Background handling

### Changed

- The exposed API uses Double values instead of Float for convenience.  

## [0.4.0](https://github.com/andreamazz/SubtleVolume/releases/tag/0.4.0)

Swift 4 support  

## [0.3.0](https://github.com/andreamazz/SubtleVolume/releases/tag/0.3.0)

Merged #11  

## [0.2.1](https://github.com/andreamazz/SubtleVolume/releases/tag/0.2.1)

Merged #10  

## [0.2.0](https://github.com/andreamazz/SubtleVolume/releases/tag/0.2.0)

Updated library to Swift 3.0

## [0.1.1](https://github.com/andreamazz/SubtleVolume/releases/tag/0.1.1)

### Fixed
- Removed observer on dealloc (See #6).

## [0.1.0](https://github.com/andreamazz/SubtleVolume/releases/tag/0.1.0)

### Added
- Carthage Support


Source/SubtleVolume.swift
//
//  SubtleVolume.swift
//  SubtleVolume
//
//  Created by Andrea Mazzini on 05/03/16.
//  Copyright © 2016 Fancy Pixel. All rights reserved.
//

import UIKit
import MediaPlayer
import AVFoundation

/**
 The style of the volume indicator

 - plain: A plain bar
 - rounded: A plain bar with rounded corners
 */
@objc public enum SubtleVolumeStyle: Int {
  case plain
  case rounded
}

/**
 The entry and exit animation of the volume indicator

 - None: The indicator is always visible
 - SlideDown: The indicator fades in/out and slides from/to the top into position
 - FadeIn: The indicator fades in and out
 */
@objc public enum SubtleVolumeAnimation: Int {
  case none
  case slideDown
  case fadeIn
}

/**
 Errors being thrown by `SubtleError`.

 - unableToChangeVolumeLevel: `SubtleVolume` was unable to change audio level
 */
public enum SubtleVolumeError: Error {
  case unableToChangeVolumeLevel
}

/**
 Delegate protocol fo `SubtleVolume`. 
 Notifies the delegate when a change is about to happen (before the entry animation)
 and when a change occurred (and the exit animation is complete)
 */
@objc public protocol SubtleVolumeDelegate {
  /**
   The volume is about to change. This is fired before performing any entry animation

   - parameter subtleVolume: The current instance of `SubtleVolume`
   - parameter value: The value of the volume (between 0 an 1.0)
   */
  @objc optional func subtleVolume(_ subtleVolume: SubtleVolume, willChange value: Double)

  /**
   The volume did change. This is fired after the exit animation is done

   - parameter subtleVolume: The current instance of `SubtleVolume`
   - parameter value: The value of the volume (between 0 an 1.0)
   */
  @objc optional func subtleVolume(_ subtleVolume: SubtleVolume, didChange value: Double)

  /**
   The volume did change. This is fired after the exit animation is done

   - parameter subtleVolume: The current instance of `SubtleVolume`
   - parameter value: The value of the volume (between 0 an 1.0)
   - returns: An optional UIImage displayed near the bar
   */
  @objc optional func subtleVolume(_ subtleVolume: SubtleVolume, accessoryFor value: Double) -> UIImage?
}

/**
 Replace the system volume popup with a more subtle way to display the volume
 when the user changes it with the volume rocker.
 */
@objc open class SubtleVolume: UIView {

  /**
   The style of the volume indicator
   */
  open fileprivate(set) var style = SubtleVolumeStyle.plain

  /**
   The entry and exit animation of the indicator. The animation is triggered by the volume
   If the animation is set to `.None`, the volume indicator is always visible
   */
  open var animation = SubtleVolumeAnimation.none {
    didSet {
      updateVolume(volumeLevel, animated: false)
    }
  }

  open var barBackgroundColor = UIColor.clear {
    didSet {
      container.backgroundColor = barBackgroundColor
    }
  }

  open var barTintColor = UIColor.white {
    didSet {
      overlay.backgroundColor = barTintColor
    }
  }

  open weak var delegate: SubtleVolumeDelegate?

  fileprivate let volume = MPVolumeView(frame: .zero)
  fileprivate let overlay = UIView(frame: .zero)
  fileprivate let container = UIView(frame: .zero)
  fileprivate let accessory = UIImageView(frame: .zero)

  /// Returns the current volume. Read-only
  public fileprivate(set) var volumeLevel = Double(0)

  /// Returns the animation state. Read-only
  public fileprivate(set) var isAnimating = false

  /// Returns the default volume bump when you programmatically change the volume
  public static let DefaultVolumeStep: Double = 0.05

  /// Padding for the inner volume bar
  public var padding: CGSize = .zero

  private var audioSessionOutputVolumeObserver: Any?


  /// Initialize with a style and a frame
  ///
  /// - Parameters:
  ///   - style: the SubtleVolumeStyle
  ///   - frame: the view's frame
  @objc convenience public init(style: SubtleVolumeStyle, frame: CGRect) {
    self.init(frame: frame)
    self.style = style
  }

  /// Initialize with a style and a frame
  ///
  /// - Parameters:
  ///   - style: the SubtleVolumeStyle
  @objc convenience public init(style: SubtleVolumeStyle) {
    self.init(style: style, frame: CGRect.zero)
  }

  @objc required public init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
    setup()
  }

  @objc required public override init(frame: CGRect) {
    super.init(frame: frame)
    setup()
  }

  required public init() {
    fatalError("Please use the convenience initializers instead")
  }


  /// Increase the volume by a given step
  ///
  /// - Parameter delta: the volume increase. The volume goes from 0 to 1, delta must be a Double in that range
  /// - Throws: `SubtleVolumeError.unableToChangeVolumeLevel`
  @objc public func increseVolume(by step: Double = DefaultVolumeStep, animated: Bool = false) throws {
    try setVolumeLevel(volumeLevel + step, animated: animated)
  }

  /// Increase the volume by a given step
  ///
  /// - Parameter delta: the volume increase. The volume goes from 0 to 1, delta must be a Double in that range
  /// - Throws: `SubtleVolumeError.unableToChangeVolumeLevel`
  @objc public func decreaseVolume(by step: Double = DefaultVolumeStep, animated: Bool = false) throws {
    try setVolumeLevel(volumeLevel - step, animated: animated)
  }

  /**
   Programatically set the volume level.

   - parameter volumeLevel: The new level of volume (between 0 an 1.0)
   - parameter animated: Indicating whether the change should be animated
   */
  @objc public func setVolumeLevel(_ volumeLevel: Double, animated: Bool = false) throws {
    guard let slider = volume.subviews.compactMap({ $0 as? UISlider }).first else {
      throw SubtleVolumeError.unableToChangeVolumeLevel
    }

    var level = volumeLevel
    if level < 0 {
      level = 0
    }
    if level > 1 {
      level = 1
    }

    let updateVolume = {
      // Trick iOS into thinking that slider value has changed
      slider.value = Float(level)
    }

    // If user opted out of animation, toggle observation for the duration of the change
    if !animated {
      audioSessionOutputVolumeObserver = nil

      updateVolume()

      DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 0.1) { [weak self] in
        guard let `self` = self else { return }
        self.audioSessionOutputVolumeObserver = AVAudioSession.sharedInstance().observe(\.outputVolume) { [weak self] (audioSession, _) in
          self?.updateVolume(level, animated: true)
        }
      }
    } else {
      updateVolume()
    }
  }

  /// Resume audio session. Call this once the app becomes active after being pushed in background
  @objc public func resume() {
    do {
      try AVAudioSession.sharedInstance().setActive(true)
    } catch {
      print("Unable to initialize AVAudioSession")
      return
    }
  }

  fileprivate func setup() {
    resume()
    updateVolume(Double(AVAudioSession.sharedInstance().outputVolume), animated: false)
    self.audioSessionOutputVolumeObserver = AVAudioSession.sharedInstance().observe(\.outputVolume) { [weak self] (audioSession, _) in
      guard let `self` = self else { return }
      self.updateVolume(Double(audioSession.outputVolume), animated: true)
    }

    backgroundColor = .clear
    container.backgroundColor = .clear

    volume.setVolumeThumbImage(UIImage(), for: UIControl.State())
    volume.isUserInteractionEnabled = false
    volume.alpha = 0.0001
    volume.showsRouteButton = false

    addSubview(volume)

    overlay.backgroundColor = .white
    container.addSubview(overlay)

    addSubview(accessory)
    addSubview(container)
  }

  open override func layoutSubviews() {
    super.layoutSubviews()

    switch style {
    case .plain:
      container.layer.cornerRadius = 0
      overlay.layer.cornerRadius = 0
    case .rounded:
      container.layer.cornerRadius = container.frame.height / 2
      overlay.layer.cornerRadius = container.frame.height / 2
    }
  }

  fileprivate func updateVolume(_ value: Double, animated: Bool) {
    delegate?.subtleVolume?(self, willChange: value)
    let previous = volumeLevel
    volumeLevel = value

    let image = delegate?.subtleVolume?(self, accessoryFor: volumeLevel)
    accessory.image = image
    var insets = UIEdgeInsets.zero
    if image != nil {
      insets.left += (frame.height)
      accessory.frame = CGRect(x: 0, y: 0, width: frame.height, height: frame.height)
    }
    insets.left += padding.width
    insets.right += padding.width
    insets.bottom += padding.height
    insets.top += padding.height

    let width = frame.size.width - insets.left - insets.right
    let height = frame.size.height - insets.top - insets.bottom

    overlay.frame = frame
    container.frame = CGRect(x: insets.left, y: insets.top, width: width, height: height)
    overlay.frame = CGRect(x: 0, y: 0, width: container.frame.width * CGFloat(previous), height: height)

    if animated {
      isAnimating = true
    }
    UIView.animate(withDuration: animated ? 0.1 : 0, animations: { () -> Void in
      self.overlay.frame.size.width = self.container.frame.width * CGFloat(self.volumeLevel)
    })

    UIView.animateKeyframes(withDuration: animated ? 2 : 0, delay: 0, options: .beginFromCurrentState, animations: { () -> Void in
      UIView.addKeyframe(withRelativeStartTime: 0, relativeDuration: 0.2, animations: {
        switch self.animation {
        case .none: break
        case .fadeIn:
          self.alpha = 1
        case .slideDown:
          self.alpha = 1
          self.transform = CGAffineTransform.identity
        }
      })

      UIView.addKeyframe(withRelativeStartTime: 0.8, relativeDuration: 0.2, animations: { () -> Void in
        switch self.animation {
        case .none: break
        case .fadeIn:
          self.alpha = 0.0001
        case .slideDown:
          self.alpha = 0.0001
          self.transform = CGAffineTransform(translationX: 0, y: -self.frame.height)
        }
      })

    }) { finished in
      self.isAnimating = !finished
      self.delegate?.subtleVolume?(self, didChange: value)
    }
  }

  deinit {
    NotificationCenter.default.removeObserver(self)
  }
}


Demo/DemoTests/DemoTests.swift
//
//  DemoTests.swift
//  DemoTests
//
//  Created by Andrea Mazzini on 05/03/16.
//  Copyright © 2016 Fancy Pixel. All rights reserved.
//

import XCTest

class DemoTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Demo/DemoUITests/DemoUITests.swift
//
//  DemoUITests.swift
//  DemoUITests
//
//  Created by Andrea Mazzini on 05/03/16.
//  Copyright © 2016 Fancy Pixel. All rights reserved.
//

import XCTest

class DemoUITests: XCTestCase {
        
    override func setUp() {
        super.setUp()
        
        // Put setup code here. This method is called before the invocation of each test method in the class.
        
        // In UI tests it is usually best to stop immediately when a failure occurs.
        continueAfterFailure = false
        // UI tests must launch the application that they test. Doing this in setup will make sure it happens for each test method.
        XCUIApplication().launch()

        // In UI tests it’s important to set the initial state - such as interface orientation - required for your tests before they run. The setUp method is a good place to do this.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // Use recording to get started writing UI tests.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
}


Demo/Demo/ViewController.swift
//
//  ViewController.swift
//  Demo
//
//  Created by Andrea Mazzini on 05/03/16.
//  Copyright © 2016 Fancy Pixel. All rights reserved.
//

import UIKit
import SubtleVolume

class ViewController: UIViewController {

  let volume = SubtleVolume(style: .rounded)
  var statusBarVisible = true

  override func viewDidLoad() {
    super.viewDidLoad()

    volume.barTintColor = .white
    volume.barBackgroundColor = UIColor.white.withAlphaComponent(0.3)
    volume.animation = .fadeIn
    volume.padding = CGSize(width: 4, height: 8)
    volume.delegate = self
    
    view.addSubview(volume)
    
    NotificationCenter.default.addObserver(volume, selector: #selector(SubtleVolume.resume), name: UIApplication.didBecomeActiveNotification, object: nil)
  }
  
  override func viewDidLayoutSubviews() {
    super.viewDidLayoutSubviews()

    if view.safeAreaInsets.top > 0 {
      volume.padding = CGSize(width: 2, height: 8)
      volume.frame = CGRect(x: 16, y: 8, width: 60, height: 20)
    } else {
      volume.frame = CGRect(x: 20, y: UIApplication.shared.statusBarFrame.height, width: UIScreen.main.bounds.width - 40, height: 20)
    }
  }

  @IBAction func minusAction() {
    do {
      try volume.decreaseVolume(animated: true)
    } catch {
      print("The demo must run on a real device, not the simulator")
    }
  }

  @IBAction func plusAction() {
    do {
      try volume.setVolumeLevel(volume.volumeLevel + 0.05, animated: true)
    } catch {
      print("The demo must run on a real device, not the simulator")
    }
  }
  
  override var preferredStatusBarStyle: UIStatusBarStyle {
    return .lightContent
  }
  
  override var preferredStatusBarUpdateAnimation: UIStatusBarAnimation {
    return .slide
  }
  
  override var prefersStatusBarHidden: Bool {
    return !statusBarVisible
  }
}

extension ViewController: SubtleVolumeDelegate {
  func subtleVolume(_ subtleVolume: SubtleVolume, accessoryFor value: Double) -> UIImage? {
    return value > 0 ? #imageLiteral(resourceName: "volume-on.pdf") : #imageLiteral(resourceName: "volume-off.pdf")
  }
  
  func subtleVolume(_ subtleVolume: SubtleVolume, didChange value: Double) {
    if !subtleVolume.isAnimating && view.safeAreaInsets.top > 0 {
      statusBarVisible = true
      UIView.animate(withDuration: 0.1) {
        self.setNeedsStatusBarAppearanceUpdate()
      }
    }
  }
  
  func subtleVolume(_ subtleVolume: SubtleVolume, willChange value: Double) {
    if !subtleVolume.isAnimating && view.safeAreaInsets.top > 0 {
      statusBarVisible = false
      UIView.animate(withDuration: 0.1) {
        self.setNeedsStatusBarAppearanceUpdate()
      }
    }
  }
}
